neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,717 @@
+from jitcdde import y as state_vector
+
+from .backend import BackendIntegrator
+from .constants import (
+    EXC,
+    MASS_NAME_STR,
+    NETWORK_CONNECTIVITY,
+    NETWORK_DELAYS,
+    NETWORK_NAME_STR,
+    NODAL_CONNECTIVITY,
+    NODAL_DELAYS,
+    NODE_NAME_STR,
+)
+from .neural_mass import NeuralMass
+
+
+class Node(BackendIntegrator):
+    """
+    Base class for all nodes within the whole-brain network.
+    """
+
+    name = ""
+    label = ""
+
+    # index of this node with respect to the whole network
+    index = None
+
+    # list of synchronisation variables that are used to compute input/output
+    # from different masses in this node, implemented as `jitcdde` helpers
+    sync_variables = []
+
+    # default network coupling - will be used if network coupling is None,
+    # usually used for isolated node
+    default_network_coupling = {}
+
+    # default output of the model - all nodes need to have this variable defined
+    default_output = None
+
+    def __init__(self, neural_masses):
+        """
+        :param neural_masses: list of neural masses in this node
+        :type neural_masses: list[NeuralMass]
+        """
+        assert all(isinstance(mass, NeuralMass) for mass in neural_masses)
+        self.masses = neural_masses
+        # number of state variables for whole node, i.e. sum of state variables
+        # for all masses
+        self.num_state_variables = sum([mass.num_state_variables for mass in self])
+        self.num_noise_variables = sum([mass.num_noise_variables for mass in self])
+        self.idx_state_var = None
+        self.initialised = False
+        assert self.default_output in self.state_variable_names[0]
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return (
+            f"Network node: {self.name} with {len(self.masses)} neural mass(es)"
+            f": {', '.join([mass.name for mass in self])}"
+        )
+
+    def describe(self):
+        """
+        Return description dict.
+        """
+        return {
+            "index": self.index,
+            "name": self.name,
+            "num_masses": len(self),
+            "num_num_state_variables": self.num_state_variables,
+            "num_noise_variables": self.num_noise_variables,
+            "masses": [mass.describe() for mass in self],
+        }
+
+    def __len__(self):
+        """
+        Get length.
+        """
+        return len(self.masses)
+
+    def __getitem__(self, index):
+        """
+        Get item.
+        """
+        return self.masses[index]
+
+    @property
+    def state_variable_names(self):
+        state_var_names = sum([mass.state_variable_names for mass in self], [])
+        # if state variables are not unique per node, append mass label to them
+        if len(state_var_names) != len(set(state_var_names)):
+            state_var_names = [
+                f"{state_var}_{mass.mass_type}" for mass in self for state_var in mass.state_variable_names
```

Comment:
no.. mass_type is mainly for distinguishing excitatory vs inhibitory masses for coupling... of course, usually, you have 1 E and 1 I mass, so it's easy... but imagine any extension of, say, W-C with 1 E but 3 I... then the local connectivity matrix and constants for coupling E-E or E-I becomes less managable...

Reply:
this particular state variables resolution of meant for this:
imagine in one node you have one EXC and one INH mass... both have firing rate as the main variable... but then when running simulation you ideally want a matrix of number of nodes x time of firing rates from EXC masses (like we do with ALN)...
this little piece of code, automatically calls firing rate from EXC `q_mean_EXC` and firing rate from INH as `q_mean_INH`... so this works when you have one node thalamus, other node AdEx, and all 4 masses have `q_mean` as the default output... then automatically you have `q_mean_EXC` for 2 masses (EXC in AdEx and TCR in thalamus) and `q_mean_INH` (INH in AdEx and TRN in thalamus)

Reply:
just now I am realising this won't work for "experimental" nodes with more than 2 masses...
imagine layer-resolving AdEx, which would be implemented as a Node with 10 masses (5 excitatory and 5 inhibitory for 5 layers)... then all excitatory masses in one big node will have the `q_mean_EXC`... then there is a question whether you want this or not... for e.g. typical FC analysis you actually want **one** output per node, this way you will have 5... so this works for classical nodes, with one exc. and one inh. masses... for large nodes, e.g. layer-resolving ones, the user needs to decide which ideally one output would be treated as an output of that large node

Reply:
That kind of goes in the direction of my question! Since masses can also have a name (later in the code `MASS_NAME_STR` is used), couldn't that just serve as the name of the state variable here as well? You'd have to predefine everything like EXC0, INH0, INH1, INH2, no?

Reply:
hmm, that's not a bad idea! firstly, I just meant name to be the name of the model kindof... but of course, one can redefine name to be something more meaningful, like `AdEx-EXC-layer1`, `AdEx-EXC-layer2` etc... then, the names can be used for definition of outputs etc... I'll try to implement it in a reasonable way)

Reply:
At implementation, this would only have to be done once and never again, right? So when I define the model, write down the equations etc, I'd give a node a name and do this 3 times in this case, all with different names, and then define the coupling matrix between masses. Soundaboutright?

Reply:
yes... my idea was to use the name and if the names are the same within one node, it will append `index` to them...
so given you have one EXC AdEx mass in the node, it will just be `AdEx_EXC`, but given you have two or more, they will become something like `AdEx_EXC_1`, `AdEx_EXC_2` etc

Reply:
I see! Don't you think though that maybe limiting the naming to EXC and INH could become a problem in the future? There are models which are EXC and INH at the same time for example. Other models might not be either of them but some other phenomenological variable that gets integrated, or multi-compartment models for example. That's why I thought it would make sense to leave this to the implementation of each model of how to call them (and not make a decision beforehand what kind of "types" there can be).

Reply:
I think I would rather keep it as general as possible: Define equations, define names, define coupling variable, define coupling matrix, done (maybe I'm missing something).

Reply:
you're missing state variable names :) but that is not a problem...

right now, it works such that each mass has defined state variable names (a list of the same length as the state vector), if you have more masses in one node with the same variable names (e.g. `q_mean` as for firing rate) then the `mass_type` is appended. Hence for e.g. ALN node this becomes `q_mean_EXC` for excitatory mass and `q_mean_INH` for inhibitory... now when you have a whole-brain network with 80 nodes, your simulation result for `q_mean_EXC` will be `80 x times` matrix and similarly, `q_mean_INH` will be `80 x time` matrix. 
I confess that this was done with only simple nodes with 1 exc. and 1 inh. node in mind... For the future, we need to think of how to do this in a very general fashion which allows automatic naming for nodes with arbitrary many masses of possible the same types... 

btw `mass_type` can be any string, so far I am using only "EXC" and "INH" because those are two mass types that I currently have, but when defining a mass or a node you can assign to `mass_type` anything... so this is not a problem when dealing with any mass type that comes into mind